### ISSUE TITLE
Feature request : add a custom block category or something to preselect some blocks 

### ISSUE THREAD

**Issue by show0k** [jmoenig/Snap--Build-Your-Own-Blocks#776]
Hi,
We are making lot of blocks to integrate [pypot](https://github.com/poppy-project/pypot) ([poppy project](https://www.poppy-project.org/?lang=fr) is an open-source software/hardware project to control robots for scientists, educators and artists) into Snap! 
The issue is that our users (many of them are not familiar with computers) are a little confused to finding blocks related to pypot in different categories. It would be nice to have an option to preselect some blocks. 
A solution could be to allow to make new categories for custom blocks.
Another solution could be to allow to export/import project with something typed in the search bar (we have added a prefix 'poppy' to all our blocks).
Have you any idea ?
If you are ok with a solution but have no time to develop it, we can make a pull request.

**Comment by jmoenig** [comment]
Yeah, I know, every hardware extension contributor wants to add their own category, so does mostly everybody else. I have - so far - felt rather strongly opposed to this, but I guess we need to maybe rethink this decision and allow user-defined categories in 4.1.

Let's keep this request open, while we discuss this internally. Thanks!

**Comment by brianharvey** [comment]
We already have the grey "other" category, in which there are no primitives, so you could put all your blocks there.  I already think we should divide variables/lists/other into three separate tabs; that would be less of a shakeup than user-defined categories.

User-defined categories are a little tricky because of zebra coloring; we have to provide block colors in light/dark pairs.  We're running out of colors!  Although I guess we could add purple like the Scratch custom-block-head color.

**Comment by show0k** [comment]
There is a grey <other> category, but they are automatically placed in the variable category, so it is not a solution.
The <pen> category is the most empty, so a hack solution could be to add them in these category.

An intermediate solution could be to make an invisible category named "hardware blocks" with a defined colour and make it visible only if there is blocks in it.

**Comment by bb010g** [comment]
I'd think that you could allow user-defined colors for new categories (perhaps use something like [HUSL](http://www.husl-colors.org/) for more accurate automatic lightness?). Or have something that generates suitable colors based off of a hash of the category name (again, HUSL may help here). The problem with using other is that it gets filled up with multiple extensions and hard to use. Subsections would help mitigate it some, but I still think the extra colors would be nice. Subsections would be nice too though for organization.

**Comment by cycomachead** [comment]
I think if "Other" were split into it's own palette it would be less necessary, but I agree it would be nice to have custom defined categories. (I hacked around a JS-function block to create a new category on the fly, but that was realllllly messy....)

If the use were mainly extensions we could have a purple category called "Extensions" ...and what I was about to type is basically this: 

> An intermediate solution could be to make an invisible category named "hardware blocks" with a defined colour and make it visible only if there is blocks in it.

So, yeah, I don't think that's too bad at all.

Actually, I think we could apply the same logic to the "Other" category -- it could have it's own palette and show it only if there are other blocks.

I also don't see a problem with custom colors, but the UI for specifying completely arbitrary categories could quickly get messy... Otherwise, we could limit to users to some small-ish of custom categories based on color. (There's a nice dark blue in BeetleBlocks that I like, and Bernant has a magenta in Snapi that I also think plays well with the current options.)

**Comment by cycomachead** [comment]
Custom categories will be a thing in v7. :)
